### DISCUSSION TITLE
Use nested target platforms with private repositories

### DISCUSSION THREAD

**Discussion by ghost** [eclipse-tycho/tycho#896] [Q&A] [answered]
Since 2.6.0, Tycho supports [nested target platforms](https://github.com/eclipse/tycho/issues/401). So it is possible to reference another target file via e.g. a file path or http address.

Our target file is hosted on a private repository, meaning that it necessary to provide credentials when trying to read this file. Otherwise Tycho fails with a "401, Unauthorized".

My initial approach was to embed both username and password into the URL via, placeholders, which are then replaced by Tycho with the actual credentials, specified in the Maven settings.
Example:
`<location type="Target" uri="https://${env_var:username}:${env_var:password}@domain.com/target-platform.target"/>`

But I'm still unable to access the file. Any suggestions?

**Comment by laeubi** [comment, marked as answer]
Tycho uses plain URL.openConnection() for this I think we can support authentication here as well. 
Just to make sure: do you see your env variables resolved in the build output (would also mean that the build output leaks your authentication!).

If you mean by "private repository" a private maven repository you could instead of https use mvn protocol to get the file instead that should handle authentication the usual way.

**Reply by ghost** [reply]
> do you see your env variables resolved in the build output

I do, which brings up a good point. Even if it would work, given that the credentials are exposed, it's probably not the solution I should go for. I also tried adding the credentials directly to the URL, skipping the string substitution, with the same effect.

> If you mean by "private repository" a private maven repository

It's an SVN repository. However, the target file is part of a Tycho build, so it wouldn't be a problem to upload it to a Maven repository.

> you could instead of https use mvn protocol

As in, add the target file as a "Maven dependency"? Hm, I can reference the artifact in the target-platform-configuration of the POM and given that Tycho supports multiple entries, that should work. 

Thank you for your response. I'll try out your suggestion and mark it as answer, if everything works fine.

**Reply by laeubi** [reply]
> As in, add the target file as a "Maven dependency"?

No you can use it as an url (just from the top of my head so you better check how exactly a target file is deployed to a GAV):

`<location type="Target" uri="mvn:[groupid]:[artifactid]:[version]:target"/>`

Beside that I think we could add authentication support for URLs I think (at least for Tycho) but then one most probably want the same for PDE, does it works there already (e.g. does a dialog poup asking for username/password)?

**Reply by ghost** [reply]
> you can use it as an url

Thanks for clarifying. That approach also works perfectly. Both in Tycho and the IDE.

> for PDE, does it works there already

No. It fails with the same 401 error.

> I think we could add authentication support for URLs.

Sounds great! Though for the time being, I think I'll stick to the Maven protocol.
